Add the ability to take a practice mode exam.

kasya opened the pull request:
- Add 'mode' field to Exam Attempt model;
 - Depending on mode show or hide correct answers during exam\practice;
 - Add tests.

kasya committed:
Add the ability to take a practice mode exam.

 - Add 'mode' field to Exam Attempt model;
 - Depending on mode show or hide correct answers during exam\practice;
 - Add tests.

arkid15r commented on apps/exam/models.py:
```diff
@@ -40,6 +40,7 @@ class ExamAttempt(models.Model):
   created = models.DateTimeField(default=timezone.now, editable=False)
   duration_minutes = models.IntegerField(default=120)
   grade = models.IntegerField(default=0)
+  mode = models.IntegerField(default=0)
```
Might be worth to consider TextField with a limited choice set

arkid15r commented on apps/exam/serializers.py:
```diff
@@ -59,8 +60,11 @@ def get_answer_attempts(self, obj):
         for obj in obj.answerattempt_set.all()
     ]
 
+  def get_mode(self, obj):
```
Not sure about this one (may be not necessary)

kasya replied:
It seems like it is necessary after all. Doesn't work with out it.

arkid15r commented on apps/exam/views.py:
```diff
@@ -50,16 +50,21 @@ def get_context_data(self, **kwargs):
     context['exam'] = exam
     context[
         'attempt_duration'] = exam.duration_minutes + self.request.user.required_extra_time
+    context['exam_mode'] = int(self.kwargs['exam_mode'])
```
The `exam` already contains  the `mode` information. There is no need to add it to the context explicitly.

kasya committed:
Resolve pull request issues.

 - Switch to TextField for exam_attempt mode;
 - Remove timer and time_left_seconds check for practice mode.

arkid15r commented on apps/exam/templates/exam/exam_intro.html:
```diff
@@ -3,7 +3,7 @@
 
 <div class="container-fluid">
-	{% if exam_mode == 1 %}
+	{% if exam_mode == "practice" %}
```
Consider creating a @property for `Attempt` class indicating the corresponding mode 

```
@property
def in_practice_mode:
  return self.mode == self.PRACTICE_MODE
```

You could also have used `exam.mode` here instead.

arkid15r commented on apps/question/views.py:
```diff
@@ -49,8 +49,11 @@ def post(self, request, **kwargs):
 
     data = request.data
     answer_ids = data.get('answers')
+    if self.attempt.mode == "exam":
```
```suggestion
    if self.attempt.in_exam_mode and not self.attempt.time_left_seconds:
      raise Http404
```

kasya committed:
Resolve pull request issues.

Switch to properties for mode in exam attempt.

kasya committed:
Show only exam mode attempts in user profile stats.

arkid15r commented on apps/user/views.py:
```diff
@@ -55,22 +55,24 @@ def get_context_data(self, **kwargs):
-        'exam', flat=True).distinct()[:show_charts]
+    exam_ids = ExamAttempt.objects.filter(
+        user=self.request.user,
+        mode='exam').values_list('exam', flat=True).distinct()[:show_charts]
```
You could use mode=ExamAttempt.EXAM_MODE here.
It seem that `.distinct()` is not really needed here and below.
Also `paginate_by = 10` and `show_charts = 2` looks weird.

arkid15r commented on apps/user/views.py:
```diff
@@ -80,8 +82,9 @@ class ProfileChart(APIView):
-    attempts = ExamAttempt.objects.filter(
-        exam=exam_id, user_id=request.user.id)[:show_entries]
+    attempts = ExamAttempt.objects.filter(exam=exam_id,
+                                          user_id=request.user.id,
```
Using `user_id=request.user_id` would be more performant.
And `user=request.user` would be more consistent.

kasya committed:
Resolve pull request issues.

arkid15r commented on apps/user/views.py:
```diff
@@ -82,9 +82,9 @@ class ProfileChart(APIView):
-    attempts = ExamAttempt.objects.filter(exam=exam_id,
-                                          user_id=request.user.id,
-                                          mode='exam')[:show_entries]
+    attempts = ExamAttempt.objects.filter(
```
Doesn't it need to be ordered  by `-created`?

kasya replied:
These are specific attempts in chart for each exam. With regular query they are sorted by created (asc). `-created` will make them look backwards. 
Also, should I specifically say order_by('created') anyway? Just in case it is taken from db in different order?

arkid15r reviewed (approved):
LGTM

kasya committed:
Resolve pull request issues.

Specifically order exam_attempts in charts by 'created' field.
